### PR TITLE
 add upload snapshot functionality using multipart request 

### PR DESCRIPTION
**Pull request by ibrahim-akrab**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [x] Have you checked your code using ```cargo clippy``` command?

This PR is related to #1567
I added the upload functionality using _actix-multipart_ with as little changes as possible. I tried not to dive in too deep and make a bunch of changes in my first pull request without consulting the mentors first.

This solution is far from optimal. Off the top of my head, I think it can be improved by not having to write the snapshot file to disk and restore it from memory. This would require some changes in the recovery logic to make it work that way (didn't look thoroughly into that but I'm willing to, if it's suitable).

Also the use of a `tempfile` then persisting it to the disk makes it look like it's writing the file twice. I checked that by tracing the system-calls it makes and apparently it just renames the old file so no additional disk load. 

I couldn't test the patch since I can't run it in distributed mode, so any help regarding that is highly appreciated.

### CODE REVIEW

File: src/actix/api/snapshot_api.rs [hunk 1]
```diff
@@ -72,6 +80,37 @@ async fn create_snapshot(
...
+    let wait = params.wait.unwrap_or(true);
+    let timing = Instant::now();
+
+    let path = format!(
+        "{}/{}/{}",
+        dispatcher.snapshots_path(),
+        collection_name,
+        snapshot.file_name.unwrap()
+    );
+    snapshot.file.persist(&path).unwrap();
```
Comment (agourlay) [inline src/actix/api/snapshot_api.rs:101]: We cannot panic in this part of the code, ideally those `unwrap` should be turned into validation errors.

**Comment by agourlay** [comment]
@ibrahim-akrab Thank you for your contribution :+1: 

I left a comment regarding dangerous unwrapping that should be addressed.

How did you test this locally? 
Ideally we would need an integration test in our [consensus tests](https://github.com/qdrant/qdrant/tree/master/tests/consensus_tests).

The test could setup a cluster, take a node down and then upload a fresh snapshot to it. 

It is a bit involved for a first contribution :)

For instance create a new test file `snapshot_upload_test.py` and run it with `pytest ./tests/consensus_tests/snapshot_upload_test.py`

File: src/actix/api/snapshot_api.rs (hunk 1 above)
Comment (ibrahim-akrab) [reply src/actix/api/snapshot_api.rs:101]: Noted. I'll fix that. 
Is there any resources on how to test it in a distributed cluster?

**Comment by ibrahim-akrab** [comment]
@agourlay Thanks for your time reviewing the PR.

I stated that I couldn't test it in distributed mode so I tested each bit in isolation. I'll fix the unwraps and definitely give consensus tests a look and write an integration test for it.

**Comment by agourlay** [comment]
Sorry for not reading your PR description properly :+1: 

Our consensus test infra. is orchestrated by `pytest` which starts several nodes locally.

One good starting for you would be [test_snapshot_recovery.py](https://github.com/qdrant/qdrant/blob/master/tests/consensus_tests/test_snapshot_recovery.py) which does something similar.

In your case, you would need to upload the snapshot instead of providing a URL to it.

**Comment by ibrahim-akrab** [comment]
@agourlay I handled the unwraps as you asked with sensible defaults when applicable (as in filenames). And I didn't know what is the maximum size a snapshot file can get since actix sets the default to 50MB and even at testing, I hit above that, so I set it to the maximum which definitely isn't ideal since it might eat up all the available space.
I also added the upload then recover test. Kindly let me know if there's anything that needs fixing.

**Comment by ibrahim-akrab** [comment]
Hey @generall, thanks for the code review. I've updated the OpenAPI definitions and generated the new ones as you described.
Regarding the maximum snapshot size limit, as I've stated before that I don't know how big it can grow so I set it initially to be as big as can get:

https://github.com/qdrant/qdrant/blob/2b7aad31144c09c59d9b5b99bfde471b5ae2793d/src/actix/mod.rs#LL91C17-L91C17

I'd hate to be introducing security risks, I have to note that this may allow a malicious actor to use up all the disk space available and I think the limit needs to be set to a reasonable size if possible since the whole point of adding this upload feature was security concerns.

**Comment by agourlay** [comment]
I understand the concern regarding the maximum snapshot size limit.

One solution is to make it configurable with a reasonable default value.

@general WDYT?

**Comment by generall** [comment]
Hi! I updated OpenAPI with missing param, I think now it should be good to go.

Regarding unlimited snapshot size - I think it should be fine. There are plenty of other methods to use up all the disk space. 
Our security concern was mostly regarding the possibility to force Qdrant to make request in the internal network.

**Review by generall — APPROVED** [review]

File: tests/consensus_tests/test_snapshot_upload.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    return r.json()["result"]["local_shards"]
+
+
+def get_remote_shards(peer_api_uri):
+    r = requests.get(f"{peer_api_uri}/collections/{COLLECTION_NAME}/cluster")
+    assert_http_ok(r)
+    return r.json()["result"]["remote_shards"]
+
+
+def recover_snapshot(peer_api_uri, snapshot_url):
```
Comment (agourlay) [inline tests/consensus_tests/test_snapshot_upload.py:39]: This function can be removed I believe

File: tests/consensus_tests/test_snapshot_upload.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        )
+        assert_http_ok(r)
+        return r.json()["result"]
+
+
+def test_upload_snapshot_1(tmp_path: pathlib.Path):
+    recover_from_uploaded_snapshot(tmp_path, 1)
+
+
+def test_upload_snapshot_2(tmp_path: pathlib.Path):
```
Comment (agourlay) [inline tests/consensus_tests/test_snapshot_upload.py:64]: this one as well

File: tests/consensus_tests/test_snapshot_upload.py (hunk 2 above)
Comment (ibrahim-akrab) [reply tests/consensus_tests/test_snapshot_upload.py:39]: Right, it's not used. Removed.

File: tests/consensus_tests/test_snapshot_upload.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        snapshot_file = {"snapshot": f}
+        r = requests.post(
+            f"{peer_api_uri}/collections/{COLLECTION_NAME}/snapshots/upload",
+            files=snapshot_file,
+        )
+        assert_http_ok(r)
+        return r.json()["result"]
+
+
+def test_upload_snapshot_1(tmp_path: pathlib.Path):
```
[the remark this answers is not in the document]
Comment (ibrahim-akrab) [reply tests/consensus_tests/test_snapshot_upload.py:60]: I saw a similar function in [`test_snaphsot_recovery.py`](https://github.com/qdrant/qdrant/blob/82a5edbe2316999827dd85476141aa5b0fa0f743/tests/consensus_tests/test_snapshot_recovery.py) and just copied the way it works over to the snapshot upload.

**Review by agourlay — APPROVED** [review]
@ibrahim-akrab LGTM, great contribution :+1:

**Comment by agourlay** [comment]
Will require the usual end touches from us:

1. write reference doc
2. integrate in drivers

**Comment by ibrahim-akrab** [comment]
@agourlay Thanks! I'd love to help in any of those final touches if possible so let me know if there's something I can do.

**Comment by agourlay** [comment]
That's very kind of you!

1. Our doc is can be found [here](https://github.com/qdrant/docs)

I 'd recommend adding the doc about snapshot update in https://github.com/qdrant/docs/blob/master/qdrant/v1.1.x/snapshots.md

2. For the drivers, I'd guide you towards the [Rust driver](https://github.com/qdrant/rust-client)

Maybe check how the download snapshot is implemented to upload using `reqwest` as well 

https://github.com/qdrant/rust-client/blob/master/src/client.rs#L1096

**Comment by agourlay** [comment]
Awesome, do not hesitate to tag me as reviewers on your PR :+1: 

Regarding the docs, you could already prepare it with only the REST example just to get started. That would be super helpful! 

The Python driver is kinda complicated to setup, here is the [dev docs](https://github.com/qdrant/qdrant_client/blob/master/tools/DEVELOPMENT.md).

You would need to modify locally the script `generate_rest_client.sh` to fetch the changes from `dev` and not `master` in order to see your changes in the models.

**Comment by ibrahim-akrab** [comment]
I actually can't find a way to represent the multipart request with the HTTP syntax used since multipart requests are usually ugly. However, `curl` can be used to send such request and it's how I have been doing it during testing. Is it okay to add that instead? It will be like the ones present in the ones in the [Quick Start Guide](https://qdrant.tech/documentation/quick_start/#create-collection)

**Comment by agourlay** [comment]
The curl syntax sounds good to me :+1:

**Comment by ibrahim-akrab** [comment]
Hey @agourlay, I'm trying to setup the python driver environment but the `generate_rest_client.sh` script fails when trying to access a non-existing repo `git@github.com:qdrant/pydantic_openapi_v3.git`. Is that a private repo or maybe an old one?

**Comment by generall** [comment]
> trying to access a non-existing repo git@github.com:qdrant/pydantic_openapi_v3.git

yes it is private. No worries, we will take care of python client ourselves

**Pull request merged**